sacred/initialize.py

```diff
@@ -354,26 +392,39 @@ def create_run(experiment, command_name, config_updates=None,
 
     # --------- configuration process -------------------
 
+
+
+
     # Phase 1: Config updates
     config_updates = config_updates or {}
+    source = CommandLineConfigSource(config_updates)
+    config_sources = {k: source for k, v in iterate_flattened(config_updates)}
     config_updates = convert_to_nested_dict(config_updates)
+    config_sources = convert_to_nested_dict(config_sources)
+
     root_logger, run_logger = initialize_logging(experiment, scaffolding,
                                                  log_level)
     distribute_config_updates(prefixes, scaffolding, config_updates)
+    distribute_config_sources(prefixes, scaffolding, config_sources)
 
     # Phase 2: Named Configs
     for ncfg in named_configs:
         scaff, cfg_name = get_scaffolding_and_config_name(ncfg, scaffolding)
         scaff.gather_fallbacks()
-        ncfg_updates = scaff.run_named_config(cfg_name)
+        ncfg_updates, ncfg_sources = scaff.run_named_config(cfg_name)
         distribute_presets(prefixes, scaffolding, ncfg_updates)
         for ncfg_key, value in iterate_flattened(ncfg_updates):
+            set_by_dotted_path(config_sources,
+                               join_paths(scaff.path, ncfg_key),
+                               ncfg_sources[ncfg_key])
             set_by_dotted_path(config_updates,
                                join_paths(scaff.path, ncfg_key),
                                value)
 
     distribute_config_updates(prefixes, scaffolding, config_updates)
 
+    distribute_config_sources(prefixes, scaffolding, config_sources)
+
```

Comment:
Wow, this piece of code got even more complicated. If this work as intendend, I am impressed. Even before your additions I was having difficulties mentally keeping track of what happens. This is not good (but not your fault). I think a serious refactoring might be needed here. 
If I understand correctly, you are capturing where a particular config entry originated from and passing this information around. Maybe there is a way to somehow encapsulate this into the config entries inside the configuration process. I'll have to think about it. (This is just me thinking out loud, not an instruction to you).

Reply:
This piece of code is very ugly and I'm not sure if it works in all cases. It worked for all scenarios I tested, but it may fail for others. I think I should rethink and rewrite this part, including tests.

Reply:
One (again very rough) idea to make it more structured and clear (I may be missing some important points and maybe it is not possible like this):

We could add these configuration origins to the ConfigSummary class and use the `ConfigSummary` throughout the whole configuration process, and not just at the end to track the changes. This would e.g. simplify the interface of `chain_evaluate_config_scopes` to just return a `ConfigSummary` and we could combine this with the config updates using a simple `ConfigSummary.update_from` (or similar). The config sources for config scopes could then be added inside the config scopes, and the config scope's `__call__` would also return a `ConfigSummary`. We could store everything hierarchically (so, every sub-`dict` would be a ConfigSummary, same for `list`s) and create an empty structure based on the ingredient hierarchy before starting the configuration process to get rid of these calls to `distribute_config_updates`, because we could directly write to the sub-`ConfigSummary` of the scaffolding using ´recursive_update´ or `set_by_dotted_path`.